File: src/bigcontext_mcp/search/tokenizer.py
"""Text tokenization for search indexing."""

import re
from dataclasses import dataclass

from bigcontext_mcp.search.stopwords import is_stop_word


@dataclass
class TokenizeOptions:
    """Options for tokenization."""

    lowercase: bool = True
    remove_stop_words: bool = True
    min_length: int = 2
    max_length: int = 50
    stemming: bool = False


def tokenize(text: str, options: TokenizeOptions | None = None) -> list[str]:
    """Tokenize text into words."""
    opts = options or TokenizeOptions()

    # Normalize text
    normalized = text
    if opts.lowercase:
        normalized = normalized.lower()

    # Remove punctuation except apostrophes in contractions
    normalized = re.sub(r"[^\w\s'-]", " ", normalized)
    normalized = re.sub(r"--+", " ", normalized)
    normalized = re.sub(r"\s+", " ", normalized)
    normalized = normalized.strip()

    # Split into words
    words = [w for w in normalized.split() if w]

    # Apply filters
    filtered_words = []
    for word in words:
        # Check length constraints
        if len(word) < opts.min_length or len(word) > opts.max_length:
            continue

        # Check if it's a number only
        if word.isdigit():
            continue

        # Remove stop words if enabled
        if opts.remove_stop_words and is_stop_word(word):
            continue

        filtered_words.append(word)

    # Apply basic stemming if enabled
    if opts.stemming:
        filtered_words = [_simple_stem(w) for w in filtered_words]

    return filtered_words


def _simple_stem(word: str) -> str:
    """Simple stemming - removes common suffixes."""
    # English suffixes
    suffixes = ["ing", "ed", "es", "s", "ment", "ness", "tion", "ation", "ly", "er", "est"]

    for suffix in suffixes:
        if word.endswith(suffix) and len(word) > len(suffix) + 2:
            return word[: -len(suffix)]

    return word


def count_term_frequencies(
    text: str, options: TokenizeOptions | None = None
) -> dict[str, int]:
    """Count term frequencies in text."""
    tokens = tokenize(text, options)
    frequencies: dict[str, int] = {}

    for token in tokens:
        frequencies[token] = frequencies.get(token, 0) + 1

    return frequencies


def get_top_terms(
    frequencies: dict[str, int], n: int
) -> list[dict[str, int | str]]:
    """Get top N terms by frequency."""
    sorted_terms = sorted(frequencies.items(), key=lambda x: x[1], reverse=True)
    return [{"term": term, "count": count} for term, count in sorted_terms[:n]]

File: src/search/tokenizer.ts
import { isStopWord } from './stopwords.js';

export interface TokenizeOptions {
  lowercase: boolean;
  removeStopWords: boolean;
  minLength: number;
  maxLength: number;
  stemming: boolean;
}

const DEFAULT_OPTIONS: TokenizeOptions = {
  lowercase: true,
  removeStopWords: true,
  minLength: 2,
  maxLength: 50,
  stemming: false,
};

/**
 * Tokenize text into words
 */
export function tokenize(text: string, options: Partial<TokenizeOptions> = {}): string[] {
  const opts = { ...DEFAULT_OPTIONS, ...options };

  // Normalize text
  let normalized = text;
  if (opts.lowercase) {
    normalized = normalized.toLowerCase();
  }

  // Remove punctuation except apostrophes in contractions
  normalized = normalized
    .replace(/[^\w\s'-]/g, ' ')
    .replace(/--+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();

  // Split into words
  let words = normalized.split(/\s+/).filter(word => word.length > 0);

  // Apply filters
  words = words.filter(word => {
    // Check length constraints
    if (word.length < opts.minLength || word.length > opts.maxLength) {
      return false;
    }

    // Check if it's a number only
    if (/^\d+$/.test(word)) {
      return false;
    }

    // Remove stop words if enabled
    if (opts.removeStopWords && isStopWord(word)) {
      return false;
    }

    return true;
  });

  // Apply basic stemming if enabled (simple suffix removal)
  if (opts.stemming) {
    words = words.map(word => simpleStem(word));
  }

  return words;
}

/**
 * Simple stemming - removes common suffixes
 * This is a basic implementation; for production use consider Porter/Snowball stemmer
 */
function simpleStem(word: string): string {
  // English suffixes
  const suffixes = ['ing', 'ed', 'es', 's', 'ment', 'ness', 'tion', 'ation', 'ly', 'er', 'est'];

  for (const suffix of suffixes) {
    if (word.endsWith(suffix) && word.length > suffix.length + 2) {
      return word.slice(0, -suffix.length);
    }
  }

  return word;
}

/**
 * Count term frequencies in text
 */
export function countTermFrequencies(text: string, options?: Partial<TokenizeOptions>): Map<string, number> {
  const tokens = tokenize(text, options);
  const frequencies = new Map<string, number>();

  for (const token of tokens) {
    frequencies.set(token, (frequencies.get(token) || 0) + 1);
  }

  return frequencies;
}

/**
 * Get top N terms by frequency
 */
export function getTopTerms(frequencies: Map<string, number>, n: number): Array<{ term: string; count: number }> {
  return Array.from(frequencies.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, n)
    .map(([term, count]) => ({ term, count }));
}
